torchdata/nodes/io/file_list.py

```diff
@@ -0,0 +1,138 @@
+import logging
+import os
+from typing import Any, Dict, List, Optional
+
+import fsspec
+from torchdata.nodes import BaseNode
+
+logger = logging.getLogger(__name__)
+
+
+class FileLister(BaseNode[Dict]):
+    """Node that lists files from any fsspec-supported filesystem matching specified patterns.
+
+    Uses fsspec to provide universal file access capabilities for local and remote filesystems.
+    Supports any filesystem that fsspec supports (local, S3, GCS, Azure, HTTP, etc.).
+
+    Features:
+    - Lists files from any fsspec-supported filesystem
+    - Supports glob patterns for file matching
+    - Maintains state for checkpointing and resumption
+
+    Input: None (configured with uri_base and patterns at initialization)
+    Output: Dictionary with file URI and metadata
+        {
+            "data": "s3://bucket/key.txt",  # File URI as string
+            "metadata": {
+                "fs_protocol": "s3",        # Filesystem protocol
+                "source": "s3"             # Source type
+            }
+        }
+
+    Examples:
+        >>> # List local files
+        >>> node = FileLister("path/to/local/dir", ["*.txt", "*.json"])
+        >>> for item in node:
+        ...     print(item["data"])  # prints "path/to/local/dir/file1.txt", etc.
+        >>>
+        >>> # List S3 files with anonymous access
+        >>> node = FileLister("s3://my-bucket/prefix", ["*.csv"],
+        ...                   fs_options={"anon": True})
+        >>> for item in node:
+        ...     print(item["data"])  # prints "s3://my-bucket/prefix/file1.csv", etc.
+        >>>
+        >>> # List GCS files
+        >>> node = FileLister("gs://my-bucket/prefix", ["*.parquet"])
+        >>> for item in node:
+        ...     print(item["data"])  # prints "gs://my-bucket/prefix/file1.parquet", etc.
+    """
+
+    SOURCE_KEY = "source"
+    CURRENT_IDX_KEY = "current_idx"
+
+    def __init__(self, uri_base: str, patterns: List[str], fs_options: Optional[Dict[str, Any]] = None):
```

Comment:
do you think it will be useful to have something like max_files.
That a person just wants to read x number of files because they might not know how many total are there